202503 Apache NuttX RTOS Contributing Guidelines update. (#15950)
apache/nuttx: CONTRIBUTING.md

# Apache NuttX RTOS Contributing Guidelines

Thank you for your interest in contributing to Apache NuttX RTOS :-)

If you haven't yet read
[The Inviolable Principles of NuttX]( https://github.com/apache/nuttx/blob/master/INVIOLABLES.md)
please do so right now.

Please note that NuttX supports over 15 different architectures, 360+ boards,
and 1600+ configurations, that are used in various project and products around
the world. Remember that any code change may affect different users and their
business. Please try to keep your contributions high quality and compatible.

> To help us smoothly process your contributions it is very important that you

Reviewer: ```suggestion
To help us process your contributions smoothly, it is very important that you

> follow the guidelines. These rules are based on daily experiences and helps

Reviewer 2: ```suggestion
follow the guidelines. These rules are based on daily experiences and help
```

> us preserve long term self-compatibility and maintenance of the project.

Reviewer: ```suggestion
us to preserve long term self-compatibility and maintenance of the project.

> NuttX is maintained by a small team of volunteers from around the world.

Reviewer 2: This is a good sentence, but I think it might need to move to elsewhere in the document, because it doesn't quite fit with the flow of this part of the text.

Author: hmm, i used top-bottom approach.. did small rearrangement.. what place would be better or what better text? :-)

This document is split into two main parts:

 1. Contribution Rules.
 2. Contribution Templates.

If you need more information please read the
[Full Contributing Documentation](https://nuttx.apache.org/docs/latest/contributing/index.html)

> or ask questions at our [Social Media channels](https://nuttx.apache.org/community).

Reviewer 2: Should we direct questions to the (archived) mailing list?

Author: I think this is shortest and most direct way to point where users may search / ask for help? Please provide text suggestions :-)

Reviewer 2: We could link to the NuttX Community page of the project website, which has mailing lists, social media, etc., and let the reader decide which to use... https://nuttx.apache.org/community/ We can leave it as is for now...

## 1. Contribution Rules.

### 1.1. Goals.

> Presented rules are here to assure high qualiy of the code and documentation,

Reviewer: ```suggestion
The rules presented are here to ensure high quality code and documentation,

standardized pull requests processing, as well as long term self-compatibilty
and maintenance of the project.

Because every change may affect users, products, or services around the world,

> rules apply equally to all authors, reviewers, committers and maintainters.

Reviewer: ```suggestion
all rules apply equally to all authors, reviewers, committers and maintainters.

This is our Check-List for processing every incoming pull request.

> Also to filter out breaking changes and handle them accordingly.

Reviewer 2: ```suggestion
Also, we filter out breaking changes and handle them accordingly.
```

### 1.2. Requirements are mandatory.

Each Contribution is a Pull Request (PR) that consists of one or more git
commits (COMMIT). Both PRs and COMMITs **must** adhere to requirements
presented in the Contributing Guidelines or will be auto-rejected

> until fixed. Special cases are possible and outlined in a dedicated rules.

Reviewer 2: ```suggestion
until fixed. Special cases are possible and outlined in dedicated rules.
```

or

```suggestion
until fixed. Special cases are possible and outlined in a separate section of dedicated rules.
```

### 1.3. Proper change description.

Proper description of change is mandatory. Description **must** contain
detailed explanation on:

 * what is the purpose of change
 * why it is necessary
 * what it does / adds / fixes
 * how things are added / changed / fixed / updated
 * what is the impact (build / runtime / api / what area)
 * how it was tested (build host, compiler, target, logs)
 * references (i.e. `nuttx` and `nuttx-apps` changes).

>  * dependencies (if change depends on other change).

Reviewer 2: ```suggestion
 * dependencies (if change depends on other changes).
```

or

```suggestion
 * dependencies (if change depends on another change).
```

Local code build and real world hardware runtime test logs
**must** be provided (for code related changes), at least short form.
For fixes, "before" and "after" logs comparison is welcome.

> Description can be single or several sentences long or bullet points

Reviewer 2: ```suggestion
Description can be a single sentence, several sentences long, or bullet points
```

> but enough for anyone to understand change goals and details.

Reviewer 2: ```suggestion
but must be enough for anyone to understand change goals and details.
```

changed, why, how), and eventual problem identification / fix in future.

### 1.4. COMMIT and PR descriptions are equally important.

Git commit messages are as important as PR descriptions.

Git log provides offline descriptions of each change that is
git client / interface independent.

### 1.5. COMMIT requirements.

Proper git commit message according to template (see 2.1) is **mandatory**,
or change is auto-rejected until fixed. Build and runtime logs are
optional here if these are too long and already provided in the PR.

Git commit message consists of:

 1. Topic with functional area prefix, `:` mark, short self-explanatory
    functional context, `.` mark.
 2. Blank line.
 3. Description on what is changed, how, and why. May use several
    lines, short sentences, or bullet points (see 1.3).
 4. Blank line.
 5. Signature (created with `git commit -s`).

Valid git commit example:

```
net/can: Add g_ prefix to can_dlc_to_len and len_to_can_dlc.

Add g_ prefix to can_dlc_to_len and len_to_can_dlc to
follow NuttX coding style conventions for global symbols,
improving code readability and maintainability.
* you can also use bullet points.

> * to note different thing briefly.

Reviewer 2: ```suggestion
* to note different things briefly.
```

Signed-off-by: AuthorName <Valid@EmailAddress>
```

### 1.6. COMMIT message mandatory fields.

Each git commit message **must** contain these fields, or change is
auto-rejected until fixed:

 1. topic.
 2. description.
 3. signature (`git commit -s`).

Although this seems to repeat rule 1.5, it clearly filters out commits
with no topic, description, or signature.

### 1.7. PR requirements.

1. Proper description (see 1.3) of PR according to template (see 2.3)
   is **mandatory**. All fields are required or change is auto-rejected
   until fixed.
2. For code changes build and runtime logs are **mandatory** to prove code
   was tested on **at least one** real world hardware target.
3. Pull Requests should be as small as possible and focused on only one
   functional change.

> 4. Different functional changes must be provided in a separate Pull Requests.

Reviewer: ```suggestion
4. Different logical functions changes must be provided in a separate Pull Requests.

Author: how about "Changes affecting different functionalities must be provided in a separate Pull Requests." ? :-)

> 4. Different functional changes must be provided in a separate Pull Requests.

Reviewer 2: ```suggestion
4. Different logical changes must be provided in a separate Pull Requests.
```

Author: "Changes affecting different functionalities must be provided in a separate  Pull Requests." ? :-)

5. PR may contain several commits but every single commit included

>    must not break break overall build, runtime, and compatibility,

Reviewer: ```suggestion
   must not break overall build, runtime, and compatibility,

   especially for other components.

> 6. PR that breaks build or runtime anyhow is considered a Breaking Change

Reviewer 2: ```suggestion
6. PR that breaks build or runtime in any way is considered a Breaking Change
```

>    (see 1.12), is not welcome, and requires special handling (see 1.13).

Reviewer 2: I don't know about "is not welcome"; how about "requires a discussion and vote of the PMC on the mailing list; see section such-and-such..."

Author: I know this sounds harsh, but its to clearly underline our approach to breaking changes, these are not welcome :-)

What others think? :-)

7. PR that introduces new feature must have Documentation included (see 1.8).
8. Work-in-progress PRs should be marked `[WIP]` tag and set as "draft".
9. When changes for a dedicated function must be bundled together in order
   to maintain functionality and self-compatibility, exception can

>    be made, and this must be clearly stated there is no other way and

Reviewer 2: ```suggestion
   be made, and it must be clearly stated that there is no other way and
```

   that this is not a Breaking Change.

> Note that small team of maintainers is processing hundreds of PRs every month.

Reviewer 2: ```suggestion
Note that our small team of maintainers is processing hundreds of PRs every month.
```

Providing high quality COMMITs and PRs really helps a lot!

Please provide only finished and verified solutions. Every PR code update
(`git push -f` to your branch that represents the PR) each time triggers big
GitHub CI machinery that builds / cross-compiles your code to various targets

> on various build hosts. This is not only slow but also expensive.

Reviewer 2: (I wonder whether draft PRs can be excluded from CI until marked as ready for review?)

Author: Yeah, that would save a lot of CI resources.. for each PR update the process is restarted and executed in full length.. maybe after first update it can be switched to manual launch mode? Maybe only syntax check on drafts? Maybe Draft by default or based on the PR template checkbox? What do you think @lupyuen ? :-)

Reviewer 3: Sorry I don't think GitHub Actions will allow us to trigger the CI Workflow when we switch from Draft Mode to Ready For Review. [(According to the docs)](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/triggering-a-workflow#using-filters)

@simbit18 Would you know if this is possible?

Actually I think it's kinda helpful for CI Workflow to run on Draft PRs. This procedure might be better for PR Review:
1. NuttX Dev submits a PR in Draft Mode
2. Wait for CI Workflow to Complete, fix any CI Errors
3. NuttX Dev sets PR to Ready for Review
4. NuttX Admins will review the PR
5. In case of fixes: NuttX Dev will revert PR to Draft Mode, wait for CI Workflow to complete, set to Ready For Review

Reviewer 4: Hi @lupyuen  and all

> Mi dispiace, ma non credo che GitHub Actions ci consentirà di attivare il flusso di lavoro CI quando passiamo dalla modalità bozza a quella pronta per la revisione. [(Secondo la documentazione)](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/triggering-a-workflow#using-filters)
> 
> @simbit18Sapresti dirmi se è possibile?

After the PR is marked as draft maybe if we check pull_request event (**github.event.pull_request.draft**) we could skip the draft PRs

Only run on pull requests that are "ready for review" not "draft" pull request.
https://github.com/reviewdog/action-eslint/issues/29#issuecomment-985939887

Don't run GitHub actions for pull requests that are drafts
https://gist.github.com/myobie/671d2f2a1e503efa0c05e7b865c26a9d

> In realtà penso che sia piuttosto utile per CI Workflow eseguire PR Draft. Questa procedura potrebbe essere migliore per PR Review:
>
> 1. NuttX Dev invia un PR in modalità bozza
> 2. Attendi il completamento del flusso di lavoro CI, correggi eventuali errori CI
> 3. NuttX Dev imposta PR su Pronto per la revisione
> 4. Gli amministratori di NuttX esamineranno il PR
> 5. In caso di correzioni: NuttX Dev ripristinerà PR in modalità bozza, attenderà il completamento del flusso di lavoro CI e lo imposterà su Pronto per la revisione

@lupyuen  Here I agree with you, also in my opinion the PR Review flow is more logical.

### 1.8. Documentation requirements.

Changes **must** come with a documentation update (where applicable).

Documentation is part of the `nuttx` git repository. If code change is part
of `nuttx-apps` repository then separate PR in `nuttx` repository is required.
Otherwise documentation should come in the same PR as the code update.

For maintenance reasons, code and documentation should be split into

> two separate commits in the same PR. This helps backporting and separates

Reviewer: @cederom I suggest remove this requirement, as I shown other project doesn't require it and even Linux that @jerpelea said requires it I looked the git history and them don't have this requirement.

@raiden00pl @xiaoxiang781216 @hartmannathan for awareness

Reviewer 5: All related change should come in one patch to avoiding the mismatch.

Reviewer 6: we will do as the majority decides

here is a Linux example
https://web.git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/log/Documentation/bpf?h=v6.14-rc5

Reviewer 7: The proposed rule reflects exactly what we've been doing for a while (same PR, separate commits). About adding it to the same (or to an individual commit), well, there is always that "gold" rule that each commit should aggregate the minimum buildable code necessary to it. That being said, this rule adheres to it: documentation is not mandatory to make the "code" buildable and should be split into different commits.

The same is valid for arch and board support: when a new feature is added, it might not be mandatory to have arch and board-code in the same commit (the arch code is buildable by itself). They can be split into different commits (in the same PR!). Documentation would be the third commit. Reverting it requires reverting all commits in the same PR.

Reviewer: @jerpelea @tmedicci I agree that is is nice to have documentation in a separated commit, but this should be suggested as good practice, but as an enforced rule. Because even projects like Linux doesn't enforce it. Never be "plus royaliste que le roi"! ;-)

Reviewer 8: There is a difference between adding something new to the documentation and updating the documentation according to changes made to the code. I don't see a problem with a separate commit for doc in the first case, but in the second case this seems strange. Documentation describes the code, if the code changes then the documentation should also be changed otherwise the doc doesn't describe the current code state.

Reviewer: > There is a difference between adding something new to the documentation and updating the documentation according to changes made to the code. I don't see a problem with a separate commit for doc in the first case, but in the second case this seems strange. Documentation describes the code, if the code changes then the documentation should also be changed otherwise the doc doesn't describe the current code state.

Exactly, it could happen very ofter: someone revert a commit and forgot the revert the other documentation commit, so the documentation will be out of sync.

possible code and documentation build errors. Squashing code together with
documentation into a single commit should be avoided.

If change presents new functionality documentation **must** be provided
along with the code (not in the future).

If change requires existing documentation update it must be contained
along with the code change (not in the future).

Successful documentation build logs (shortcut) are welcome.

This helps us keep documentation in sync with the code.

See:
1. https://github.com/apache/nuttx/tree/master/Documentation
2. https://github.com/apache/nuttx/blob/master/INVIOLABLES.md

### 1.9. Zero Trust approach to user testing.

> We implement zero trust approach to user provided testing.

Reviewer 2: ```suggestion
We implement zero trust approach to user-provided testing.
```

> It is the change author duty to provide real world hardware build and runtime

Reviewer 2: ```suggestion
It is the change author's duty to provide real world hardware build and runtime
```

breaks build / runtime compatibility with existing release code (packages)
both for nuttx and nuttx-apps, etc.

Remember that any code change may impact other users and their business!

Changes should provide maximum self-compatibility with existing solutions,
should not impact build and runtime in a negative way, or may require design
reconsiderations such as providing alternative non-invasive user selectable
solutions (see 1.10).

When breaking changes are unavoidable, these may be accepted, but need prior
discussion and agreement of the community and special handling (see 1.13).

See: https://github.com/apache/nuttx/blob/master/INVIOLABLES.md

### TODO: 1.13. Breaking changes handling process.

This rule complements "Breaking changes not welcome" rule.

We avoid breaking changes unless absolutely necessary and unavoidable
(i.e. security fix, broken code, etc). Special case considerations apply:

 1. First reviewer that recognizes a breaking change should block
    accidental merge with "Request Changes" mark and ask for discussion.
    This PR cannot be merged until all requests are resolved.
 2. PR is marked as "Draft" to avoid accidental merge.

>  3. Detailed discussion should follow both in PR **and** dev@ Mailing List.

Reviewer 7: Should we provide a reference for it?
```suggestion
 3. Detailed discussion should follow both in PR **and** [dev@ Mailing List](https://nuttx.apache.org/community/#mailing-list).
```

 * 2 or more independent positive reviews.
 * 4 or more independent positive reviews for Breaking Changes (see 1.12).

In future we plan to add "areas" (such as trivial typo fixes or documentation
only updates) that may require only one reviewer.

### 1.16. Reviews independence.

PR reviews should come from independent organizations.
NuttX is an Open-Source independent and unbiased project.

Each PMC Member, Committer, and Reviewer must report up-to-date Affiliation
for clear identification. Commits should contain business email where
applicable.

When code comes from the same organization as positive review, then review
is not considered independent, but still may provide helpful insight.

Self review is not allowed.

See: https://github.com/apache/nuttx/blob/master/INVIOLABLES.md

### 1.17. Merge rules.

1. Each change **must** be provided as PR that undergoes independent review
   process (see 1.16).
2. Single company / organization commit, review, and merge is not allowed.
3. Merge of PRs with unresolved discussions and "change request" marks
   is not allowed.
3. Self committed code merge with or without review is not allowed.

> 4. Direct push to master is not allowed.

Author: Definitely :-) I am kinda new here but this is not allowed on most repos that I know, not sure why it was enabled, but there may be a reason? We will switch it after rules are applied for sure :-)

If you haven't yet read
[The Inviolable Principles of NuttX]( https://github.com/apache/nuttx/blob/master/INVIOLABLES.md)
please do so first.

Please note that NuttX supports over 15 different architectures, 360+ boards,
and 1600+ configurations, that are used in various projects and products around
the world. Remember that any code change may affect different users and their
businesses. Please try to keep your contributions high quality and compatible.

To help us process your contributions smoothly, it is very important that you
follow the guidelines. NuttX is maintained by a small team of volunteers from
around the world. These rules are based on daily experiences and help us
to preserve long term self-compatibility and maintenance of the project.

This document is split into two main parts:

 1. Contribution Rules.
 2. Contribution Templates.

If you need more information please read the
[Full Contributing Documentation](https://nuttx.apache.org/docs/latest/contributing/index.html)
or ask questions at our [Social Media channels](https://nuttx.apache.org/community).

## 1. Contribution Rules.

### 1.1. Goals.

> The rules presented are here to ensure high qualiy code and documentation,

Reviewer 2: ```suggestion
The rules presented are here to ensure high quality code and documentation,
```

Author: thank you fixed! :-)

> standardized pull requests processing, as well as long term self-compatibilty

Reviewer 2: ```suggestion
standardized pull requests processing, as well as long term self-compatibility
```

Author: thank you fixed! :-)
